File: backend/app/schemas/procedure.py
"""
Procedure & Intervention schemas.

Phase 9: Schemas for flexible procedure tracking across all specialties.
"""

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.procedure import ProcedureOutcome, ProcedureSeverity


class ProcedureBase(BaseModel):
    """Base procedure schema."""

    category: str = Field(..., min_length=1, max_length=100)
    procedure_type: str = Field(..., min_length=1, max_length=100)
    sub_type: str | None = Field(None, max_length=100)
    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None


class ProcedureCreate(ProcedureBase):
    """Schema for creating a procedure."""

    patient_id: UUID
    doctor_id: UUID
    appointment_id: UUID | None = None

    # Timing
    performed_at: datetime | None = None  # Defaults to now
    duration_minutes: int | None = Field(None, ge=1, le=1440)

    # Outcome
    outcome: ProcedureOutcome = ProcedureOutcome.SUCCESSFUL
    severity: ProcedureSeverity = ProcedureSeverity.MINOR

    # Clinical Details
    findings: str | None = None
    notes: str | None = None
    complications: str | None = None

    # Consumables (flexible JSON for specialty-specific data)
    # Examples:
    # Cardiology: {"stent_brand": "Xience", "stent_size": "3.0x18mm", "quantity": 2}
    # Orthopedics: {"implant_type": "Hip", "manufacturer": "Zimmer"}
    consumables: dict[str, Any] | None = None

    # Custom fields for specialty-specific measurements
    # Examples:
    # Cardiology: {"ef_before": 45, "ef_after": 55}
    # Ophthalmology: {"preop_vision": "6/36", "postop_vision": "6/6"}
    custom_fields: dict[str, Any] | None = None

    # Billing
    billing_code: str | None = Field(None, max_length=20)
    cpt_code: str | None = Field(None, max_length=20)
    icd_code: str | None = Field(None, max_length=20)
    is_billable: bool = True
    billed_amount: Decimal | None = Field(None, ge=0)

    # Location & Team
    location: str | None = Field(None, max_length=100)
    assistant_doctors: list[str] | None = None

    # Follow-up
    requires_followup: bool = False
    followup_notes: str | None = None


class ProcedureUpdate(BaseModel):
    """Schema for updating a procedure."""

    category: str | None = Field(None, max_length=100)
    procedure_type: str | None = Field(None, max_length=100)
    sub_type: str | None = Field(None, max_length=100)
    name: str | None = Field(None, max_length=255)
    description: str | None = None

    performed_at: datetime | None = None
    duration_minutes: int | None = Field(None, ge=1, le=1440)

    outcome: ProcedureOutcome | None = None
    severity: ProcedureSeverity | None = None

    findings: str | None = None
    notes: str | None = None
    complications: str | None = None

    consumables: dict[str, Any] | None = None
    custom_fields: dict[str, Any] | None = None

    billing_code: str | None = None
    cpt_code: str | None = None
    icd_code: str | None = None
    is_billable: bool | None = None
    billed_amount: Decimal | None = None

    location: str | None = None
    assistant_doctors: list[str] | None = None

    requires_followup: bool | None = None
    followup_notes: str | None = None


class ProcedureResponse(BaseModel):
    """Schema for procedure response."""

    id: UUID
    patient_id: UUID
    doctor_id: UUID
    appointment_id: UUID | None
    clinic_id: UUID

    category: str
    procedure_type: str
    sub_type: str | None
    name: str
    description: str | None

    performed_at: datetime
    duration_minutes: int | None

    outcome: str
    severity: str

    findings: str | None
    notes: str | None
    complications: str | None

    consumables: dict[str, Any] | None
    custom_fields: dict[str, Any] | None

    billing_code: str | None
    cpt_code: str | None
    icd_code: str | None
    is_billable: bool
    billed_amount: float | None

    location: str | None
    assistant_doctors: list[str] | None

    requires_followup: bool
    followup_notes: str | None

    emr_procedure_id: str | None
    synced_to_emr: bool

    created_at: datetime
    updated_at: datetime

    # Nested details (populated when needed)
    patient_name: str | None = None
    doctor_name: str | None = None

    model_config = {"from_attributes": True}


class ProcedureListResponse(BaseModel):
    """Schema for listing procedures."""

    id: UUID
    patient_id: UUID
    patient_name: str
    doctor_id: UUID
    doctor_name: str
    category: str
    procedure_type: str
    name: str
    performed_at: datetime
    outcome: str
    severity: str
    is_billable: bool
    billed_amount: float | None

    model_config = {"from_attributes": True}


# ========== Analytics Schemas ==========

class ProcedureStats(BaseModel):
    """Procedure statistics for a time period."""

    total: int
    by_category: dict[str, int]
    by_type: dict[str, int]
    by_outcome: dict[str, int]
    by_severity: dict[str, int]
    total_billed: float
    average_duration_minutes: float | None


class ProcedureTypeCount(BaseModel):
    """Count of procedures by type."""

    category: str
    procedure_type: str
    count: int
    total_billed: float


class DoctorProcedureStats(BaseModel):
    """Procedure statistics per doctor."""

    doctor_id: UUID
    doctor_name: str
    total_procedures: int
    by_category: dict[str, int]
    total_billed: float
    success_rate: float


class ProcedureTrend(BaseModel):
    """Daily procedure trend."""

    date: datetime
    count: int
    categories: dict[str, int]
    total_billed: float


# ========== Template Schemas ==========

class ProcedureTypeTemplate(BaseModel):
    """Template for a procedure type."""

    procedure_type: str
    sub_types: list[str] | None = None
    default_duration: int | None = None
    default_billing_code: str | None = None
    required_fields: list[str] | None = None


class ProcedureCategoryTemplate(BaseModel):
    """Template for a procedure category."""

    category: str
    types: dict[str, list[str]]


class ProcedureTemplatesResponse(BaseModel):
    """Response with all procedure templates."""

    templates: dict[str, ProcedureCategoryTemplate]


# ========== Quick Log Schemas ==========

class ProcedureQuickLog(BaseModel):
    """Schema for quick procedure logging (minimal fields)."""

    patient_id: UUID
    category: str
    procedure_type: str
    name: str
    outcome: ProcedureOutcome = ProcedureOutcome.SUCCESSFUL
    notes: str | None = None
    consumables: dict[str, Any] | None = None
    billed_amount: Decimal | None = None


class CardiacProcedureLog(BaseModel):
    """Specialized schema for cardiac procedures."""

    patient_id: UUID
    procedure_type: str  # Echo, Angioplasty, Stent, etc.
    name: str

    # Cardiac-specific fields
    ef_before: float | None = Field(None, ge=0, le=100)
    ef_after: float | None = Field(None, ge=0, le=100)
    lesion_location: str | None = None  # LAD, LCX, RCA, etc.

    # Stent-specific
    stent_brand: str | None = None
    stent_size: str | None = None
    stent_quantity: int | None = Field(None, ge=1)

    outcome: ProcedureOutcome = ProcedureOutcome.SUCCESSFUL
    complications: str | None = None
    notes: str | None = None
    billed_amount: Decimal | None = None


class OphthalmologyProcedureLog(BaseModel):
    """Specialized schema for ophthalmology procedures."""

    patient_id: UUID
    procedure_type: str  # Cataract, LASIK, Injection, etc.
    name: str
    eye: str = Field(..., pattern="^(left|right|both)$")

    # Vision-specific
    preop_vision: str | None = None  # e.g., "6/36"
    postop_vision: str | None = None  # e.g., "6/6"

    # Lens-specific
    lens_type: str | None = None
    lens_power: str | None = None

    outcome: ProcedureOutcome = ProcedureOutcome.SUCCESSFUL
    complications: str | None = None
    notes: str | None = None
    billed_amount: Decimal | None = None

File: mobile/lib/core/models/procedure.dart
/// Procedure model for tracking medical procedures
///
/// Phase 9: Procedure & Intervention Tracking
class Procedure {
  final String id;
  final String patientId;
  final String doctorId;
  final String? appointmentId;
  final String clinicId;
  final String category; // Cardiology, Orthopedics, etc.
  final String procedureType; // Echo, Stent, Surgery
  final String? subType;
  final DateTime procedureDate;
  final Map<String, dynamic>? consumables;
  final Map<String, dynamic>? customFields;
  final String? icdCode;
  final String? cptCode;
  final String outcome; // successful, partial, failed, referred
  final String? notes;
  final double? billedAmount;
  final String? invoiceId;
  final DateTime createdAt;
  final DateTime updatedAt;

  // Nested data
  final String? patientName;
  final String? doctorName;

  Procedure({
    required this.id,
    required this.patientId,
    required this.doctorId,
    this.appointmentId,
    required this.clinicId,
    required this.category,
    required this.procedureType,
    this.subType,
    required this.procedureDate,
    this.consumables,
    this.customFields,
    this.icdCode,
    this.cptCode,
    required this.outcome,
    this.notes,
    this.billedAmount,
    this.invoiceId,
    required this.createdAt,
    required this.updatedAt,
    this.patientName,
    this.doctorName,
  });

  factory Procedure.fromJson(Map<String, dynamic> json) {
    return Procedure(
      id: json['id'] as String,
      patientId: json['patient_id'] as String,
      doctorId: json['doctor_id'] as String,
      appointmentId: json['appointment_id'] as String?,
      clinicId: json['clinic_id'] as String,
      category: json['category'] as String,
      procedureType: json['procedure_type'] as String,
      subType: json['sub_type'] as String?,
      procedureDate: DateTime.parse(json['procedure_date'] as String),
      consumables: json['consumables'] as Map<String, dynamic>?,
      customFields: json['custom_fields'] as Map<String, dynamic>?,
      icdCode: json['icd_code'] as String?,
      cptCode: json['cpt_code'] as String?,
      outcome: json['outcome'] as String,
      notes: json['notes'] as String?,
      billedAmount: json['billed_amount'] != null
          ? (json['billed_amount'] as num).toDouble()
          : null,
      invoiceId: json['invoice_id'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      patientName: json['patient_name'] as String?,
      doctorName: json['doctor_name'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'patient_id': patientId,
      'doctor_id': doctorId,
      'appointment_id': appointmentId,
      'clinic_id': clinicId,
      'category': category,
      'procedure_type': procedureType,
      'sub_type': subType,
      'procedure_date': procedureDate.toIso8601String().split('T')[0],
      'consumables': consumables,
      'custom_fields': customFields,
      'icd_code': icdCode,
      'cpt_code': cptCode,
      'outcome': outcome,
      'notes': notes,
      'billed_amount': billedAmount,
      'invoice_id': invoiceId,
    };
  }
}

/// Procedure statistics
class ProcedureStats {
  final int totalProcedures;
  final int successfulCount;
  final int partialCount;
  final int failedCount;
  final int referredCount;
  final double totalBilled;
  final double successRate;

  ProcedureStats({
    required this.totalProcedures,
    required this.successfulCount,
    required this.partialCount,
    required this.failedCount,
    required this.referredCount,
    required this.totalBilled,
    required this.successRate,
  });

  factory ProcedureStats.fromJson(Map<String, dynamic> json) {
    return ProcedureStats(
      totalProcedures: json['total_procedures'] as int,
      successfulCount: json['successful_count'] as int,
      partialCount: json['partial_count'] as int,
      failedCount: json['failed_count'] as int,
      referredCount: json['referred_count'] as int,
      totalBilled: (json['total_billed'] as num).toDouble(),
      successRate: (json['success_rate'] as num).toDouble(),
    );
  }
}

/// Procedure type count for analytics
class ProcedureTypeCount {
  final String category;
  final String procedureType;
  final int count;
  final double totalBilled;

  ProcedureTypeCount({
    required this.category,
    required this.procedureType,
    required this.count,
    required this.totalBilled,
  });

  factory ProcedureTypeCount.fromJson(Map<String, dynamic> json) {
    return ProcedureTypeCount(
      category: json['category'] as String,
      procedureType: json['procedure_type'] as String,
      count: json['count'] as int,
      totalBilled: (json['total_billed'] as num).toDouble(),
    );
  }
}

/// Procedure template for quick logging
class ProcedureTemplate {
  final String category;
  final List<String> types;
  final List<String> subtypes;
  final List<String> commonConsumables;

  ProcedureTemplate({
    required this.category,
    required this.types,
    required this.subtypes,
    required this.commonConsumables,
  });

  factory ProcedureTemplate.fromJson(Map<String, dynamic> json) {
    return ProcedureTemplate(
      category: json['category'] as String,
      types: List<String>.from(json['types'] as List),
      subtypes: List<String>.from(json['subtypes'] as List),
      commonConsumables: List<String>.from(json['common_consumables'] as List),
    );
  }
}
